### PR TITLE
[Train] implement CheckpointStrategy

### PR DESCRIPTION
**Pull request by matthewdeng**
Implements `CheckpointStrategy` to delete checkpoints from disk.

## Why are these changes needed?
For a training function with many epochs, persisted checkpoints may lead to a high disk usage. The user should be able to define in their training run to clean up checkpoints. Specifically, they may want to keep N checkpoints, removing checkpoints based on:
1. Timestamp - remove old and outdated checkpoints.
2. A user-defined metric - remove "bad" or low-scoring checkpoints.

### Example
This example is also included in the docs.
```python
from ray import train
from ray.train import CheckpointStrategy, Trainer


def train_func():
    # first checkpoint
    train.save_checkpoint(loss=2)
    # second checkpoint
    train.save_checkpoint(loss=4)
    # third checkpoint
    train.save_checkpoint(loss=1)
    # fourth checkpoint
    train.save_checkpoint(loss=3)

# Keep the 2 checkpoints with the smallest "loss" value.
checkpoint_strategy = CheckpointStrategy(num_to_keep=2,
                                         checkpoint_score_attribute="loss",
                                         checkpoint_score_order="min")

trainer = Trainer(backend="torch", num_workers=2)
trainer.start()
trainer.run(train_func, checkpoint_strategy=checkpoint_strategy)
print(trainer.best_checkpoint_path)
# /home/ray_results/train_2021-09-01_12-00-00/run_001/checkpoints/checkpoint_000003
print(trainer.latest_checkpoint_dir)
# /home/ray_results/train_2021-09-01_12-00-00/run_001/checkpoints
print([checkpoint_path for checkpoint_path in trainer.latest_checkpoint_dir.iterdir()])
# [PosixPath('/home/ray_results/train_2021-09-01_12-00-00/run_001/checkpoints/checkpoint_000003'),
# PosixPath('/home/ray_results/train_2021-09-01_12-00-00/run_001/checkpoints/checkpoint_000001')]
trainer.shutdown()
```

## Implementation
The implementation loosely follows that of Tune's [checkpoint_manager](https://github.com/ray-project/ray/blob/master/python/ray/tune/checkpoint_manager.py).

1. Add a `_top_persisted_checkpoints` priority queue in `CheckpointManager` to keep track of the "best" checkpoints. Checkpoint paths are wrapped in a `PersistedCheckpoint` object.
2. The user can customize checkpoint retention logic in `CheckpointStrategy`. This will determine when and which checkpoints to delete. **Note:** If two checkpoints have the same priority, the _original_ one will be kept and the _latest_ one will never be persisted.
3. Store a single `_best_persisted_checkpoint`, which is exposed in `Trainer.best_checkpoint_path`. This replaces `Trainer.last_checkpoint_path`, as the most chronologically recent persistent checkpoint path may be already deleted based on the `CheckpointStrategy`. The default strategy uses timestamp, which would allow this field to effectively remain the same.
4. A default `_timestamp` attribute is added to checkpoints for the purpose of supporting timestamp based retention.

## Checks

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: python/ray/util/sgd/v2/backends/backend.py [hunk 1]
```diff
@@ -121,18 +143,84 @@ def _load_checkpoint(self,
...
+        checkpoint_score_attribute = \
+            self._checkpoint_strategy.checkpoint_score_attribute
+        checkpoint_score_order = \
+            self._checkpoint_strategy.checkpoint_score_order
+        if checkpoint_score_attribute not in checkpoint:
+            raise ValueError(f"Unable to persist checkpoint for "
+                             f"checkpoint_score_attribute: "
+                             f"{checkpoint_score_attribute}. "
+                             f"Include this attribute in the call to "
+                             f"save_checkpoint.")
```
Comment (amogkam) [inline python/ray/util/sgd/v2/backends/backend.py:161]:
```suggestion
                             f"sgd.save_checkpoint().")
```

File: python/ray/util/sgd/v2/backends/backend.py [hunk 2]
```diff
@@ -121,18 +143,84 @@ def _load_checkpoint(self,
...
+        checkpoint_score = checkpoint[checkpoint_score_attribute]
+        try:
+            checkpoint_score = float(checkpoint_score)
```
Comment (amogkam) [inline python/ray/util/sgd/v2/backends/backend.py:164]: You might want to use `numbers.Numbers` to check instead of trying to cast to float.

https://stackoverflow.com/questions/29773143/how-to-determine-that-a-python-object-is-numeric

File: python/ray/util/sgd/v2/backends/backend.py [hunk 3]
```diff
@@ -40,6 +41,18 @@ class SGDBackendError(Exception):
...
+        self.priority = priority
+
+    def __lt__(self, other):
+        return self.priority < other.priority
+
+    def __repr__(self):
+        return f"PersistedCheckpoint({repr(self.path)})"
+
+
 class CheckpointManager:
```
Comment (amogkam) [inline python/ray/util/sgd/v2/backends/backend.py:56]: Should we move this and `PersistentCheckpoint` into the `checkpoint` module? This is starting to contain more logic so maybe we want to separate them from `backend`.

File: python/ray/util/sgd/v2/backends/backend.py [hunk 4]
```diff
@@ -157,6 +245,14 @@ def latest_checkpoint_path(self) -> Optional[Path]:
         else:
```
Comment (amogkam) [inline python/ray/util/sgd/v2/backends/backend.py:245]: Can we consolidate the implementation for `latest_checkpoint_path` and `best_checkpoint_path`? Currently best_checkpoint_path uses the PersistedCheckpoint object to get the path, but latest_checkpoint_path does path manipulation to get the path given the directory and file name. I think we can have both of these use the path stored in the `PersistedCheckpoint` object.

Comment (amogkam) [reply python/ray/util/sgd/v2/backends/backend.py:245]: Actually do we even need `latest_checkpoint_path` anymore?

File: python/ray/util/sgd/v2/tests/test_trainer.py [hunk 5]
```diff
@@ -439,28 +439,111 @@ def test_persisted_checkpoint(ray_start_2_cpus, logdir):
...
+
+    def validate():
+        checkpoint = sgd.load_checkpoint()
+        assert checkpoint is not None
+        assert checkpoint["loss"] == 3
+
+    trainer.run(validate, checkpoint=trainer.best_checkpoint_path)
+
+
+def test_persisted_checkpoint_strategy_failure(ray_start_2_cpus):
```
Comment (amogkam) [inline python/ray/util/sgd/v2/tests/test_trainer.py:522]: This is already being tested in the previous test right?

File: python/ray/util/sgd/v2/trainer.py [hunk 6]
```diff
@@ -361,14 +361,17 @@ def latest_checkpoint_dir(self) -> Optional[Path]:
         return self._executor.latest_checkpoint_dir
 
     @property
-    def latest_checkpoint_path(self) -> Optional[Path]:
-        """Path to the latest persisted checkpoint from the latest run.
+    def best_checkpoint_path(self) -> Optional[Path]:
```
Comment (amogkam) [inline python/ray/util/sgd/v2/trainer.py:364]: Do we also want to expose a `best_checkpoint` object the same way we have for `latest_checkpoint`?

**Review by amogkam — COMMENTED** [review]
Nice! Is there docs for this?

**Review by amogkam — APPROVED** [review]
LGTM! Just left some minor documentation points
